src/Compilers/CSharp/Portable/Binder/ForEachLoopBinder.cs

```diff
@@ -277,7 +277,7 @@ private BoundForEachStatement BindForEachPartsWorker(DiagnosticBag diagnostics,
                             hasErrors = true;
                         }
 
-                        boundIterationVariableType = new BoundTypeExpression(variables, aliasOpt: null, type: iterationVariableType);
+                        boundIterationVariableType = new BoundTypeExpression(variables, aliasOpt: null, type: iterationVariableType).MakeCompilerGenerated();
```

Comment:
>boundIterationVariableType [](start = 24, length = 26)

As we discussed, bound types are supposed to reflect types in source; you're filing a follow-up issue to separate the representation of the type in source from the type of the iterator.

Reply:
Filed https://github.com/dotnet/roslyn/issues/24105 from our discussion. Thanks